// bambu-js/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertState = exports.hms = exports.errors = exports.printStageValues = exports.PrintStage = exports.BambuPrinter = void 0;
const BambuPrinter_1 = __importDefault(require("./classes/BambuPrinter"));
exports.BambuPrinter = BambuPrinter_1.default;
const PrintStage_1 = require("./types/PrintStage");
Object.defineProperty(exports, "PrintStage", { enumerable: true, get: function () { return PrintStage_1.PrintStage; } });
Object.defineProperty(exports, "printStageValues", { enumerable: true, get: function () { return PrintStage_1.printStageValues; } });
const stateConverter_1 = __importDefault(require("./utilities/stateConverter"));
exports.convertState = stateConverter_1.default;
const errors = __importStar(require("./utilities/errors"));
exports.errors = errors;
const hms = __importStar(require("./utilities/hms"));
exports.hms = hms;
exports.default = BambuPrinter_1.default;
